File: frondosa-nursery-shop/source/scripts/source-frondosa-images.py
#!/usr/bin/env python3
import json, re, time, urllib.parse, urllib.request
from pathlib import Path

OUT = Path('assets')
UA = 'NeutronFrondosaImageRepair/1.0 (local portfolio demo; source attribution manifest)'
TARGETS = {
  'birdNestFern': ['Asplenium nidus bird nest fern potted', 'Asplenium nidus'],
  'jade': ['Crassula ovata jade plant potted', 'Crassula ovata'],
  'hoya': ['Hoya carnosa hanging plant', 'Hoya carnosa'],
  'pilea': ['Pilea peperomioides potted plant', 'Pilea peperomioides'],
  'orchid': ['Phalaenopsis orchid potted flower', 'Phalaenopsis orchid'],
  'aloe': ['Aloe vera potted plant', 'Aloe vera'],
  'cachepot': ['ceramic cachepot flower pot', 'ceramic flower pot planter', 'flower pot ceramic'],
  'terracottaSaucer': ['terracotta flower pot saucer', 'terracotta pot'],
  'heatPack': ['disposable hand warmer heat pack', 'hand warmer packet', 'shipping heat pack'],
}

def req(url):
    r = urllib.request.Request(url, headers={'User-Agent': UA})
    return urllib.request.urlopen(r, timeout=40)

def search(q):
    params = {'action':'query','format':'json','generator':'search','gsrsearch':q,'gsrnamespace':'6','gsrlimit':'10','prop':'imageinfo','iiprop':'url|mime|size|extmetadata','iiurlwidth':'1200'}
    with req('https://commons.wikimedia.org/w/api.php?' + urllib.parse.urlencode(params)) as f:
        data=json.load(f)
    pages=sorted(data.get('query',{}).get('pages',{}).values(), key=lambda p:p.get('index',99))
    c=[]
    for p in pages:
        info=(p.get('imageinfo') or [{}])[0]
        if info.get('url') and re.match(r'image/(jpeg|png|webp)', info.get('mime','')) and info.get('width',0)>=350 and info.get('height',0)>=350:
            c.append((p['title'], info))
    return c

def filename(key,title,mime):
    ext={'image/png':'png','image/webp':'webp'}.get(mime,'jpg')
    base=re.sub(r'\.[^.]+$','',title.replace('File:',''))
    base=re.sub(r'[^A-Za-z0-9]+','-',base).strip('-').lower()[:86]
    return f'{key}-{base}.{ext}'

OUT.mkdir(exist_ok=True)
manifest_path=OUT/'image-sources.json'
try:
    manifest=json.loads(manifest_path.read_text())
except Exception:
    manifest={'generatedAt': time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime()), 'source':'Wikimedia Commons API imageinfo; chosen to replace repeated/misidentified Frondosa catalog primary images.', 'images':{}}

for key, queries in TARGETS.items():
    if key in manifest['images'] and (OUT/manifest['images'][key]['filename']).exists():
        print('skip', key)
        continue
    chosen=None
    for q in queries:
        time.sleep(4)
        try:
            cs=search(q)
        except Exception as e:
            print('search fail', key, q, e)
            time.sleep(20); continue
        if cs:
            chosen=(q, cs[0][0], cs[0][1]); break
    if not chosen:
        print('no candidate', key); continue
    q,title,info=chosen
    name=filename(key,title,info['mime'])
    url=(info.get('thumburl') or info['url']).split('?')[0]
    data = None
    for attempt in range(5):
        try:
            time.sleep(8*(attempt+1))
            with req(url) as f:
                data = f.read()
            (OUT/name).write_bytes(data)
            break
        except Exception as e:
            print('download fail', key, attempt+1, e)
    if data is None:
        continue
    ext=info.get('extmetadata') or {}
    manifest['images'][key]={'query':q,'filename':name,'commonsTitle':title,'descriptionUrl':info.get('descriptionurl'),'width':info.get('width'),'height':info.get('height'),'bytes':len(data),'artist':re.sub('<[^>]+>','',ext.get('Artist',{}).get('value','')),'licenseShortName':ext.get('LicenseShortName',{}).get('value',''),'usageTerms':ext.get('UsageTerms',{}).get('value','')}
    manifest_path.write_text(json.dumps(manifest, indent=2)+'\n')
    print(key, name, len(data))
print('manifest', manifest_path)

File: frondosa-nursery-shop/source/scripts/source-frondosa-images.mjs
import fs from 'node:fs/promises';
import path from 'node:path';

const outDir = path.resolve('assets');
const manifestPath = path.join(outDir, 'image-sources.json');
const USER_AGENT = 'NeutronFrondosaImageRepair/1.0 (local portfolio demo; source attribution manifest)';
const wait = (ms) => new Promise(resolve => setTimeout(resolve, ms));

const targets = [
  { key: 'zz', queries: ['Zamioculcas zamiifolia potted plant', 'Zamioculcas zamiifolia houseplant'] },
  { key: 'parlorPalm', queries: ['Chamaedorea elegans potted plant', 'Chamaedorea elegans'] },
  { key: 'rubber', queries: ['Ficus elastica potted', 'Ficus elastica plant'] },
  { key: 'stringHearts', queries: ['Ceropegia woodii String of Hearts hanging plant', 'Ceropegia woodii'] },
  { key: 'birdNestFern', queries: ['Asplenium nidus bird nest fern potted', 'Asplenium nidus'] },
  { key: 'jade', queries: ['Crassula ovata jade plant potted', 'Crassula ovata'] },
  { key: 'hoya', queries: ['Hoya carnosa hanging plant', 'Hoya carnosa'] },
  { key: 'pilea', queries: ['Pilea peperomioides potted plant', 'Pilea peperomioides'] },
  { key: 'orchid', queries: ['Phalaenopsis orchid potted flower', 'Phalaenopsis orchid'] },
  { key: 'aloe', queries: ['Aloe vera potted plant', 'Aloe vera'] },
  { key: 'cachepot', queries: ['ceramic cachepot flower pot', 'ceramic flower pot planter', 'flower pot ceramic'] },
  { key: 'terracottaSaucer', queries: ['terracotta pot saucer', 'terracotta flower pot saucer', 'terracotta pot'] },
  { key: 'heatPack', queries: ['disposable hand warmer heat pack', 'hand warmer packet', 'shipping heat pack'] },
];

function safeName(key, title, ext) {
  const cleaned = title.replace(/^File:/, '').replace(/\.[^.]+$/, '').normalize('NFKD').replace(/[^a-zA-Z0-9]+/g, '-').replace(/^-|-$/g, '').slice(0, 86).toLowerCase();
  return `${key}-${cleaned}.${ext}`;
}

async function fetchJson(url, tries = 4) {
  for (let i = 0; i < tries; i++) {
    const res = await fetch(url, { headers: { 'User-Agent': USER_AGENT } });
    if (res.ok) return res.json();
    if (![429, 500, 502, 503, 504].includes(res.status) || i === tries - 1) throw new Error(`HTTP ${res.status}: ${url}`);
    await wait(1200 * (i + 1));
  }
}

async function commonsSearch(query) {
  const params = new URLSearchParams({
    action: 'query', format: 'json', generator: 'search', gsrsearch: query,
    gsrnamespace: '6', gsrlimit: '10', prop: 'imageinfo', iiprop: 'url|mime|size|extmetadata'
  });
  const data = await fetchJson(`https://commons.wikimedia.org/w/api.php?${params}`);
  const pages = Object.values(data.query?.pages ?? {}).sort((a,b) => (a.index ?? 99) - (b.index ?? 99));
  return pages
    .map(p => ({ title: p.title, ...(p.imageinfo?.[0] ?? {}) }))
    .filter(p => p.url && /^image\/(jpeg|png|webp)$/.test(p.mime || '') && p.width >= 350 && p.height >= 350);
}

async function download(url, filename, tries = 5) {
  for (let i = 0; i < tries; i++) {
    const res = await fetch(url, { headers: { 'User-Agent': USER_AGENT } });
    if (res.ok) {
      const buf = Buffer.from(await res.arrayBuffer());
      await fs.writeFile(path.join(outDir, filename), buf);
      return buf.length;
    }
    if (![429, 500, 502, 503, 504].includes(res.status) || i === tries - 1) throw new Error(`download ${res.status}: ${url}`);
    await wait(2000 * (i + 1));
  }
}

await fs.mkdir(outDir, { recursive: true });
const manifest = { generatedAt: new Date().toISOString(), source: 'Wikimedia Commons API imageinfo; chosen to replace repeated/misidentified Frondosa catalog primary images.', images: {} };

for (const target of targets) {
  let chosen, queryUsed;
  for (const query of target.queries) {
    await wait(350);
    const candidates = await commonsSearch(query);
    if (candidates.length) { chosen = candidates[0]; queryUsed = query; break; }
  }
  if (!chosen) throw new Error(`No Commons image candidates for ${target.key}: ${target.queries.join(' | ')}`);
  const ext = chosen.mime === 'image/png' ? 'png' : chosen.mime === 'image/webp' ? 'webp' : 'jpg';
  const filename = safeName(target.key, chosen.title, ext);
  const bytes = await download(chosen.url, filename);
  manifest.images[target.key] = {
    query: queryUsed,
    filename,
    commonsTitle: chosen.title,
    descriptionUrl: chosen.descriptionurl,
    width: chosen.width,
    height: chosen.height,
    bytes,
    artist: chosen.extmetadata?.Artist?.value?.replace(/<[^>]+>/g, '') ?? '',
    licenseShortName: chosen.extmetadata?.LicenseShortName?.value ?? '',
    usageTerms: chosen.extmetadata?.UsageTerms?.value ?? '',
  };
  console.log(`${target.key}: ${filename} (${chosen.width}x${chosen.height}, ${bytes} bytes)`);
}
await fs.writeFile(manifestPath, JSON.stringify(manifest, null, 2) + '\n');
console.log(`manifest: ${manifestPath}`);
